### PR TITLE
Ballerina Check-pointing Initial Implementation

### PR DESCRIPTION
**Pull request by warunalakshitha**
## Purpose
 > Ballerina checkpointing provides the ability to persist the state of an execution, so the execution can be resumed from the checkpointed state in the future. 

```ballerina
service<http:Service> s1 bind { port: 9090 } {
    
	@http:ResourceConfig {
        methods:["POST"],
        path:"/r1",
	body: "userDetails"
    }
    @Interruptible
    r1 (endpoint conn, http:Request req, json userDetails) {
    
        string userid =userDetails.userid.toString()
	Person p1 = new;
	p1.name = "smith";
	p1.age = 20;
	runtime:checkpoint();
	runtime:sleep(5000);	                
	http:Response res = new;
	_ = conn -> respond(res);
	io:println("Process completed.");

    }
}
```
If the system is interrupted when at the sleep, execution can be resumed from the checkpointed state once the server recovered next time.

## Approach

This is supported by persisting the state in the filesystem as a JSON. When saving the state, the context hierarchy is persisted. This includes worker data as well as global and local properties of each context. All primitive types and non-reference types are persisted. Currently, only BMap and BRefArray are persisted from reference types. 

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

## Automation tests
 - Unit tests 
   > Unit tests provided.
 - Integration tests 
   > Integration tests provided.

### CODE REVIEW

File: bvm/ballerina-core/src/main/java/org/ballerinalang/model/InterruptibleNativeCallableUnit.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.ballerinalang.model;
+
+/**
+ * Interface represents interruptible @{@link NativeCallableUnit}. This is used to determine the persistence of
+ * @{@link NativeCallableUnit}.
+ *
+ * @since 0.976.0
```
Comment (MaryamZi) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/model/InterruptibleNativeCallableUnit.java]: version needs to be updated?

File: bvm/ballerina-core/src/main/java/org/ballerinalang/connector/impl/ResourceExecutor.java [hunk 2]
```diff
@@ -59,6 +62,13 @@ public static void execute(Resource resource, CallableUnitCallback responseCallb
         }
         ResourceInfo resourceInfo = resource.getResourceInfo();
         if (properties != null) {
+            Object o = properties.get(Constants.IS_INTERRUPTIBLE);
+            if (o != null && (boolean) o) {
```
Comment (Kishanthan) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/connector/impl/ResourceExecutor.java]: use proper variable names here. "o" is not a proper name.

File: bvm/ballerina-core/src/main/java/org/ballerinalang/model/InterruptibleNativeCallableUnit.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+/**
+ * Interface represents interruptible @{@link NativeCallableUnit}. This is used to determine the persistence of
+ * @{@link NativeCallableUnit}.
+ *
+ * @since 0.981.1
+ *
+ */
+public interface InterruptibleNativeCallableUnit extends NativeCallableUnit {
+
+    boolean persistBeforeOperation();
```
Comment (Kishanthan) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/model/InterruptibleNativeCallableUnit.java]: Lets add proper java doc for the API methods.

File: bvm/ballerina-core/src/main/java/org/ballerinalang/persistence/Deserializer.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+/**
+ * Class manages the deserialization functionality.
+ *
+ * @since 0.981.1
+ */
+public class Deserializer {
+
+    // These data structures will be used store deserialize objects temporary to manage sharable objects.
+    private Map<String, BRefType> tempRefTypes = new ConcurrentHashMap<>();
```
Comment (Kishanthan) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/persistence/Deserializer.java]: Why these variable names are starting with temp* ? Shall we just use without temp*?

**Comment by warunalakshitha** [comment]
@Kishanthan yes. change @interruptible to lowercase to comply with other built-in annotations.

**Pull request merged**